DOC: Teams analytics events and SQL tables

catong opened the pull request:
https://openedx.atlassian.net/browse/DOC-1980
This PR adds documentation for teams-related analytics events only, described in this wiki page: https://openedx.atlassian.net/wiki/display/AN/Teams+Feature+Event+Design

New SQL tables created for Teams (https://openedx.atlassian.net/browse/TNL-1965) will be documented in a separate PR: https://github.com/edx/edx-documentation/pull/555

This feature will be in next week's release, Sept 9. Documentation will not be published on docs.edx.org, but will be shared with PMs and course teams for the teams pilot.

### Reviewers
Check the appropriate box after each reviewer finishes and gives :+1:.
- [x] Subject matter expert: @dianakhuang  
- [x] Doc team review: @lamagnifica 
- [x] Product review: @explorerleslie 
- [ ] PM review: Alison Kuryla
- [x] Analytics team review: @stroilova 

### Draft HTML Documentation
- [x] http://draft-research-guide-for-teams.readthedocs.org/en/latest/internal_data_formats/tracking_logs.html#student-teams-events
- [x] http://draft-research-guide-for-teams.readthedocs.org/en/latest/internal_data_formats/sql_schema.html#teams-courseteam
- [x] See also http://draft-research-guide-for-teams.readthedocs.org/en/latest/internal_data_formats/tracking_logs.html#discussion-forums-events where "team_id" was added to the forum events.

### Post-review
- [x] Squash commits

dianakhuang commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+       long, the string is truncated, ``...`` is added at the end of the
+       string, and this field is included in the ``truncated`` array.
+
+.. What is the char limit for the old and new fields?
```
This should be the same as the limit for the `user_api` events. By default, it's 1250 characters, I believe.

dianakhuang commented:
That was the only comment I had. Any other questions/clarifications I could make?

lamagnifica commented on en_us/data/source/internal_data_formats/event_list.rst:
```diff
@@ -111,6 +111,20 @@ D, E, F
+     - :ref:`student_teams_events`
+   * - ``edx.team.learner_removed``
+     - :ref:`student_teams_events`
+   * - ``edx.team.changed``
```
move to line 114 to present alphabetically

lamagnifica commented on en_us/data/source/internal_data_formats/event_list.rst:
```diff
@@ -111,6 +111,20 @@ D, E, F
+     - :ref:`student_teams_events`
+   * - ``edx.team.searched``
+     - :ref:`student_teams_events`
+   * - ``edx.team.page_viewed``
```
move to line 124 to alphabetize

lamagnifica commented on en_us/data/source/internal_data_formats/event_list.rst:
```diff
@@ -111,6 +111,20 @@ D, E, F
+     - :ref:`student_teams_events`
    * - ``edx.problem.hint.demandhint_displayed``
      - :ref:`problem`
    * - ``edx.problem.hint.feedback_displayed``
```
move entire set of team events to start on line 132, after the edx.problem... events

catong replied:
Argh. Alphabetization FAIL!

catong replied:
Fixed, and also moved the "library events" to the "L" position

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -2994,6 +2994,11 @@ purpose for comments as they do for threads or responses.
+     - string
+     - If the comment was part of a team discussion within a course, this
+       field shows the identifier of the team that the comment was created in. For
+       more information about teams events, see :ref:`student_teams_events`.
```
teams events --> events for teams?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -2994,6 +2994,11 @@ purpose for comments as they do for threads or responses.
      - dictionary
      - Contains a member ``id`` field with the unique identifier of the
        response that the user added this comment to.
+   * - ``team_id``
```
I think this could be described once, at the thread level, and added to line 2981 and 3028

catong commented:
Hi @dianakhuang. I'm going through the eventing PR (https://github.com/edx/edx-platform/pull/9607) to verify the possible values for fields such as "add_method" and "remove_method", so if you could review those in particular that would be appreciated. Thanks!

dianakhuang commented:
@catong There are three possible values for each, and I'm just going to list them here for your convenience:

add_method  
* `added_on_create`
* `joined_from_team_view`
* `added_by_another_user`

remove_method
* `self_removal`
* `removed_by_admin`
* `team_deleted`

catong commented:
Thanks, @dianakhuang you're awesome! For add_method, are "added_by_staff" and "invite_link" (which are on the wiki page) valid options, but only for future functionality?

dianakhuang commented:
@catong `invite_link` is a possibility in the future, but I believe `added_by_staff` is not going to be used.

catong commented:
OK. Thanks for the clarification!

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3206,6 +3216,13 @@ complete, the server emits an ``edx.forum.thread.created`` event.
        
+    * - ``team_id``
+     - string
+     - If the thread was part of a team discussion within a course, this field
```
this is kind of awkward. Maybe simpler to say,
If this thread is part of a team discussion within a course, this field identifies the team.
or something like
Discussions can be private to the members of a team of learners within a course. This field identifies the team involved in a private discussion.

catong replied:
I simplified, but "private" is not accurate because team discussions can be viewed by everyone.

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3206,6 +3216,13 @@ complete, the server emits an ``edx.forum.thread.created`` event.
+     - string
+     - If the thread was part of a team discussion within a course, this field
+       shows the identifier of the team that the thread was created in. For
+       more information about teams events, see :ref:`student_teams_events`.
```
teams events --> events for teams of students?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+.. _student_teams_events:
+
+==========================
+Student Teams Events
```
"teams events" sounds weird to me

catong replied:
@lamagnifica I tried out "Student Team Events" but that sounded even weirder to me, like a school sports meet. "Student Teams Events" is parallel with some of our other headings, such as "Discussion Forums Events"... even though we don't call them forums anymore!  Do you have any suggestions? Maybe I'll go with "Student Team Events" despite weirdness, as it's parallel with, and appears directly after, "Student Cohort Events".

catong replied:
Actually, decided to stick with "Student Teams Events". We refer to Course Team Events; I think it might get confusing if we have "Student Team Events". "Teams" plural sounds more like a feature.

lamagnifica replied:
what about Teams-Related Events? you use that later in descriptions

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+* ``edx.team.changed``
+* ``edx.team.searched``
+* ``edx.team.page_viewed``
+
```
moving toward presenting events in alpha order in this rst also (slowly, over time)

lamagnifica replied:
this might also be a good place to note that some of these events are actually course team events, and not student events, but that they are all presented here for convenience (?). Or, move the events that only a Staff or Admin role-holder can do to the course team events section later in this file

catong replied:
Uh-oh: "Course Team Team Events" ;-p

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``team_id``
+     - string
+     - The identifier for the team.
+
```
@dkh is there a SQL table that gets populated with these teams IDs? if so, do you know if we include it in the data packages?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - Field
+     - Type
+     - Details
+   * - ``team_id``
```
rather than provide this info repeatedly, I suggest describing it once, such as for the created event, and then mentioning it with an xref to the created event. 
For example, the load_video event includes this info after the ``event`` **Member Fields** header:

The load_video... events include the following event member fields. These fields serve the same purpose for events of this type as for the play_video... events.

http://edx.readthedocs.org/projects/devdata/en/latest/internal_data_formats/tracking_logs.html#load-video-edx-video-loaded

- code
- id

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+     - The identifier for the team.
+   * - ``user_id``
+     - string
+     - The identifier for the learner who joined or was added to the team.  
```
@dkh can you confirm that this is a string? Other events have descriptions that identify this ID as an integer, such as edx.cohort.user_added: "The numeric ID (from auth_user.id) of the added user."

http://edx.readthedocs.org/projects/devdata/en/latest/internal_data_formats/tracking_logs.html#load-video-edx-video-loaded

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``add_method``
+     - string
+     - The method by which the learner joined the team. For example,
+       ``joined_from_team_view``, ``added_by_staff``, or ``invite_link``.
```
are these the only possible values, or are there others? If possible, list all of the possible values, and not a subset as am example

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``remove_method``
+     - string
+     - The method by which the learner was removed from the team. For example,
+       ``self_removal``, ``team_deleted``, or ``removed_by_admin``.  
```
ditto about listing all possible values

catong replied:
That is all values, but I'll change the wording from "For example" so it doesn't sound like this is not the exhaustive list.

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+     - The identifier for the team.
+   * - ``user_id``
+     - string
+     - The identifier for the learner who left or was removed from the team.
```
ditto on type and learner questions

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+     - Details
+   * - ``field``
+     - string
+     - The name of the field that was modified.
```
an xref to the doc that describes what can be changed would be helpful here, or a list of the fields that could possibly be changed.

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``new``
+     - string
+     - The value of the field after the modification. If this value is too
+       long, the string is truncated, ``...`` is added at the end of the
```
what is too long? 2000 characters?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3921,271 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+
+
+   * - ``truncated``
+     - string
```
string or array? (if this is a string, I suggest changing array to field 2x above)

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+     - string
+     - If the string representation of the old or new value of a field is
+       longer than 1250 characters, the string is truncated and ``...`` is
+       added at the end. The ``truncated`` event field is an array of the
```
The first sentence describes the possible effect on the old and new fields, not the truncated field.

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+       end of the string, and this field is included in the ``truncated``
+       array.
+   * - ``truncated``
+     - string
```
is this a string or an array?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+       longer than 1250 characters, the string is truncated and ``...`` is
+       added at the end. The ``truncated`` event field is an array of the
+       ``old`` and ``new`` fields that have been truncated.
+
```
the only other truncated field in events now works quite differently from this (see link below), so I'm just guessing here, but if this is an array, I suspect it contains an "old" member field with a value, possibly the complete string, or the part that got truncated, or a "new" field with the same kind of value, or both "old" and "new" (if the field that changed had a super long value that got changed to another super long value)
@dianakhuang a sample event would help a lot for this one.

http://edx.readthedocs.org/projects/devdata/en/latest/internal_data_formats/tracking_logs.html#edx-forum-thread-created

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+``edx.team.team_searched``
+*********************************
+
+When a search for teams has been performed from the topic view under the
```
change to active voice if possible

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``topic_id``
+     - string
+     - The identifier for the topic under which this search for teams was
+       performed.
```
this description doesn't give enough context -- possibly xref to the description of how you use topics to search for teams?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+``edx.team.page_viewed``
+*********************************
+
+When any page with a unique URL under the **Teams** page of the Instructor
```
this seems really specific -- what might have a unique URL under the Teams page that is something other than a page with team info on it?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``page_name``  
+     - string
+     - The name of the page that was viewed. For example, ``all-topics``,
+       ``single-team``, ``my-teams``.
```
list all values if possible rather than an example.

catong replied:
Changed wording to indicate this is a list of all possible values.

dianakhuang replied:
@catong I don't believe this is a complete list of all possible page names. I can get a list for you.

dianakhuang replied:
Currently, we have:
* `my-teams`
* `browse`
* `single-topic`
* `single-team`
* `new-team`
* `edit-team`
* `search-teams`

catong replied:
Thanks @dianakhuang !

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+     - string
+     - The identifier of the topic related to the page that was viewed. This
+       value is set to ``null`` if the topic is not applicable or does not
+       exist.
```
yeah, the topics thing is hard to understand without any context

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``team_id``
+     - string
+     - The identifier of the team related to the page that was viewed. This
+       value is set to ``null`` if the team is not applicable or does not
```
what is the scenario results in a team not being applicable?
could it be restated as, 
If the page does not provide info about a team, this field is set to ``null``. (?)

lamagnifica commented:
editing pass complete. Several questions and suggestions.

explorerleslie commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+*********************************
+
+When a team is deleted, the server emits an ``edx.team.deleted`` event. Only a
+course team member who has the **Staff** or **Admin** role can delete teams.
```
@catong I'm pretty sure we expanded this to include anyone that had any of the Discussion roles as well (Admin, Moderator, TA). @dianakhuang can you please confirm? 

Do we call out that global staff (edX internal employees mostly on edx.org) can also perform these actions, or is it implied that anyone who has global staff has pretty much every permission?

dianakhuang replied:
Yes, TAs and moderators also have the power to delete teams as well.

catong replied:
Updated, thanks

explorerleslie commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+*********************************
+
+When a learner joins a team or is added by someone else, the server emits an
+``edx.team.learner_added`` event.
```
@catong RIght now the only way for a user to be added to a team is by joining themselves, unless there is a direct API call made. Should we say just "when a learner joins a team" or do we also account for API calls made outside the UI in our event descriptions?

catong replied:
I would err on the side of a wider description, not limited to only events triggered in the UI. @lamagnifica is this what we have done?

explorerleslie commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3920,269 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+     - The count of results that matched the search text.  
+   * - ``search_text``
+     - string
+     - The text or keywords that were used in the search.
```
Might be a good place to use active voice to avoid confusion "the text that the user used in the search" (although the weird "user used" phrase would be introduced...)

explorerleslie commented:
@catong a few minor questions / comments, otherwise :+1: I don't need to re-review before merge unless there are substantial changes.

catong commented:
@lamagnifica @dianakhuang @explorerleslie if you want to take a last look at the events doc. Changes since previous review mostly organizational, e.g. extracting out common events and cross-referencing one place rather than repeating info in many events.
FYI the SQL table doc will be in a separate PR.

catong commented:
Squashed commits

dianakhuang commented:
Other than that, it looks good. :+1:

catong commented:
@stroilova Do you want to take a look? Thanks!

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3125,26 +3134,24 @@ complete, the server emits an ``edx.forum.thread.created`` event.
      - Applies only to courses that allow discussion posts that are anonymous
-       to all other users.
-       
+       to all other users.       
```
the extra lines were in here and below to get the second sentence to appear as a separate paragraph.

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3918,298 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+Student Teams Events
+==========================
+
+This section includes descriptions of the following events, generated when a
```
events, generated when a --> events, which are generated if a

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3918,298 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+teams-related actions. This list includes both student events and course team
+events, because some of these events are triggered by actions that can be
+performed both by students and course staff, or by students with special roles
+such as community TAs as well as course staff.
```
what do you mean by "course staff" here? course team members with the Staff role? or course team members in general, either Admin or Staff?
the only other special role that a student might have is Beta Tester -- list them both, or eliminate "such as" depending on whether being a bets tester has any relevance to this feature?
our roles and their interactions are so confusing!

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3918,298 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+*Building and Running an edX Course* guide.
+
+
+.. _edx_team_team_changed:
```
is the name of this event edx.team.changed or edx.team.team_changed?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3918,298 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+This section presents teams-related events alphabetically. Typically, the
+first event produced when teams are included in a course is the
+``edx.team.created`` event.
+
```
is there a reason not to use contents:: here?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3918,298 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+
+**Event Source**: Server
+
+**History** Added 9 Sept 2015.
```
update throughout? or will there be pilot course events out there starting from this date?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3918,298 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+   * - ``truncated``
+     - array
+     - The ``truncated`` event field is an array of the ``old`` and ``new``
+       fields that have been truncated.
```
I'm still not understanding what this field does. 
@dkh, would you just let me know if I've got this scenario right?
I'm on a team and I change the value in the Description field from "{1500 characters}" to "{5 characters}". An event is emitted. 
The field in the event is Description. 
The new in the event is "{5 characters}". 
The old in the event is "{1250 characters}...". 
The truncated in the event is [old].
The event does not provide the last 250 characters from the old description anywhere.
Thanks, I appreciate your time.

dianakhuang replied:
@lamagnifica Yes, that's how it should behave. We actually do something similar in the `user_api` events, so maybe if we reuse that language it's less confusing?

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3918,298 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+     - The identifier of the topic related to the page that was viewed. This
+       value is set to ``null`` if a topic is not applicable to the page that
+       was viewed, or if the topic does not exist.
+   * - ``team_id``
```
is there a reason to include this up here as well as on line 4169?

catong replied:
Ack. I meant to delete after copying.

lamagnifica commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -4811,3 +5121,10 @@ uploading a .csv file of student cohort assignments.
+.. note that the above Teams Course Author doc is available only in draft
+.. form. As of 15 Sept 2015 the feature used in a course pilot but not
+.. widely released on edx.org, so doc has not been published. When it is
+.. published we need to update the link to point to docs.edx.org location.
```
would you please add a JIRA story and include its number here?

catong commented:
@dianakhuang I've incorporated the list of pages for edx.team.page_viewed, and added the new event: edx.team.activity_updated.

dianakhuang commented:
@catong Looks good to me :+1:

lamagnifica commented:
thanks, @catong . :+1: 
PS Do NOT click "Update branch" below until we know better what's going on with GitHub

catong commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3922,335 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+
+.. note:: The Teams feature is in limited release. For more information,
+   contact your edX Partner Manager. For Open edX sites, contact your system
+   administrator.
```
@explorerleslie. Same note text as in the SQL tables sections.

stroilova commented on en_us/data/source/internal_data_formats/tracking_logs.rst:
```diff
@@ -3904,6 +3922,335 @@ Instructor Dashboard, the server emits an ``edx.cohort.user_removed`` event.
+events and course team events, because some of these events are triggered by
+actions that can be performed both by students and course staff (with the
+Staff, Admin, Discussion Admin or Discussion Moderator roles), or by students
+with special roles such as community TAs.
```
Minor q: Is "community" here usually capitalized?

catong replied:
Sharp eyes! Thank you for catching that @stroilova

stroilova commented:
:+1:

catong commented:
Brought in SQL tables documentation that was originally in https://github.com/edx/edx-documentation/pull/555

catong commented:
Merging https://github.com/edx/edx-documentation/pull/535
